plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/state/MapboxViewportStateBase.kt

```diff
@@ -0,0 +1,156 @@
+package com.mapbox.maps.plugin.viewport.experimental.state
+
+import android.animation.Animator
+import android.animation.AnimatorSet
+import android.animation.ValueAnimator
+import androidx.core.animation.doOnEnd
+import com.mapbox.maps.plugin.animation.CameraAnimationsPlugin
+import com.mapbox.maps.plugin.viewport.DEFAULT_FRAME_TRANSITION_MAX_DURATION_MS
+import com.mapbox.maps.plugin.viewport.experimental.ViewportPlugin
+import com.mapbox.maps.plugin.viewport.experimental.data.ViewportData
+import com.mapbox.maps.plugin.viewport.experimental.data.ViewportDataSource
+import com.mapbox.maps.plugin.viewport.experimental.data.ViewportDataSourceUpdateObserver
+import com.mapbox.maps.plugin.viewport.experimental.transition.TransitionEndListener
+import com.mapbox.maps.plugin.viewport.experimental.transition.ViewportStateTransition
+import com.mapbox.maps.plugin.viewport.experimental.transition.ViewportTransitionOptions
+import java.util.concurrent.CopyOnWriteArrayList
+
+abstract class MapboxViewportStateBase : ViewportState {
+  protected var frameTransitionOptions =
+    ViewportTransitionOptions.Builder().maxDuration(DEFAULT_FRAME_TRANSITION_MAX_DURATION_MS)
+      .build()
+  protected var runningAnimation: AnimatorSet? = null
+  protected val transitionEndListeners = CopyOnWriteArrayList<TransitionEndListener>()
+
+  abstract val stateTransition: ViewportStateTransition
+  abstract val cameraPlugin: CameraAnimationsPlugin
+
+  abstract val dataSource: ViewportDataSource
+
+  private val dataSourceUpdateObserver =
+    ViewportDataSourceUpdateObserver { viewportData ->
+      if (!isTransitioning) {
+        updateFrame(viewportData, true)
+      }
+    }
+
+  /**
+   * Boolean flag that indicate if it is in a transition state from previous state.
+   */
+  override var isTransitioning: Boolean = false
+    protected set
+
+  /**
+   * Start transition to this state.
+   *
+   * @param stateTransitionOptions defines the transition options for the state transition from previous state.
+   * @param frameTransitionOptions defines the transition options for the continuous frame updates after the state transition.
+   * @param transitionEndListener listener that will be called after the state transition has finished.
+   */
+  override fun transitionToState(
+    stateTransitionOptions: ViewportTransitionOptions,
+    frameTransitionOptions: ViewportTransitionOptions,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    dataSource.registerUpdateObserver(dataSourceUpdateObserver)
```

Comment:
If the observer delivers the currently available data synchronously upon registration, this would initiate a frame transition before the actual state transition.

Reply:
This part of the codebase has been refactored, please review again.